pulsar-zookeeper-utils/src/main/java/org/apache/pulsar/zookeeper/ZooKeeperCache.java

```diff
@@ -296,32 +297,57 @@ public Boolean call() throws Exception {
         CompletableFuture<Optional<Entry<T, Stat>>> future = new CompletableFuture<>();
         dataCache.get(path, (p, executor) -> {
             // Return a future for the z-node to be fetched from ZK
+            final CompletableFuture<Boolean> existsFuture = new CompletableFuture<>();
             CompletableFuture<Entry<Object, Stat>> zkFuture = new CompletableFuture<>();
 
             // Broker doesn't restart on global-zk session lost: so handling unexpected exception
             try {
-                this.zkSession.get().getData(path, watcher, (rc, path1, ctx, content, stat) -> {
-                    Executor exec = scheduledExecutor != null ? scheduledExecutor : executor;
+                // Check for existence first, then get data: this allows us to put the watch on unconditionally.
+                zkSession.get().exists(path, watcher, (rc, path1, ctx, stat) -> {
```

Comment:
In most of the cases, we try to get data for nodes that exist so, what if we try to get the data first and if not present then we trigger `exists(..)` to set the watch?